### PR TITLE
Introduce new image library, Phase I.

### PR DESCRIPTION
**Pull request by prideout**
The preps for our upcoming mipgen command line tool by adding filtering
capabilities and unit tests for a new LinearImage class, which will
subsume Image (deprecated).

Filament's new image library is composed of three components:

- LinearImage....simple 3D tensor of floats
- ImageOps.......free functions for simple transforms
- ImageSampler...high quality filtering

The old Image class was a simple untyped data blob and is now
deprecated. The new LinearImage class is always float32, and always
stores pixels in packed scanline order. This makes it easier to
implement image-based algorithms because they can be agnostic of the
underlying format.

Remaining refactorings:
- Phase  II will migrate imageio to LinearImage
- Phase III will migrate all downstream tools to LinearImage
- Phase  IV will remove old the Image class

### CODE REVIEW

File: libs/image/CMakeLists.txt [hunk 1]
```diff
@@ -40,3 +43,11 @@ target_compile_options(${TARGET} PRIVATE
 # ==================================================================================================
 add_library(image_headers INTERFACE)
 target_include_directories(image_headers INTERFACE ${PUBLIC_HDR_DIR})
+
+# ==================================================================================================
+# Tests
+# ==================================================================================================
+add_executable(test_${TARGET}
```
Comment (romainguy) [inline libs/image/CMakeLists.txt]: You need to compile this only on host targets, not Android.

File: libs/image/include/image/ColorTransform.h [hunk 2]
```diff
@@ -198,6 +199,65 @@ std::unique_ptr<uint8_t[]> fromLinearToRGBM(const Image& image) {
     return dst;
 }
 
+template <typename T>
+std::unique_ptr<uint8_t[]> fromLinearTosRGB(const LinearImage& image) {
+    using math::float3;
+    size_t w = image.getWidth();
+    size_t h = image.getHeight();
+    size_t channels = image.getChannels();
```
Comment (romainguy) [inline libs/image/include/image/ColorTransform.h]: There's a mix of hard-coded `3` and `channels`.

File: libs/image/include/image/ImageOps.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#define IMAGE_IMAGEOPS_H
+
+#include <image/LinearImage.h>
+
+#include <initializer_list>
+
+namespace image {
+
+// Concatenates images horizontally to create a filmstrip atlas, similar to numpy's hstack.
+LinearImage hstack(std::initializer_list<LinearImage> images);
```
Comment (romainguy) [inline libs/image/include/image/ImageOps.h]: Don't abbreviate :)

File: libs/image/include/image/LinearImage.h [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    uint32_t getWidth() const { return mWidth; }
+    uint32_t getHeight() const { return mHeight; }
+    uint32_t getChannels() const { return mChannels; }
+    void reset() { *this = LinearImage(); }
+
+private:
+
+    // Pixel data has shared ownership to facilitate shallow copies and a convenient Python-like API
+    // that allows clients to easily pass around images.
+    std::shared_ptr<float> mData;
```
Comment (romainguy) [inline libs/image/include/image/LinearImage.h]: No STL in public headers

File: libs/image/src/ImageSampler.cpp [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+#include <utils/Panic.h>
+
+#include <memory>
+#include <vector>
+
+using namespace image;
+
+namespace {
+
+struct FilterFn {
```
Comment (romainguy) [inline libs/image/src/ImageSampler.cpp]: `FilterFunction`?

File: libs/image/CMakeLists.txt (hunk 1 above)
Comment (prideout) [reply libs/image/CMakeLists.txt]: Done

File: libs/image/include/image/ColorTransform.h (hunk 2 above)
Comment (prideout) [reply libs/image/include/image/ColorTransform.h]: Done

File: libs/image/include/image/ImageOps.h (hunk 3 above)
Comment (prideout) [reply libs/image/include/image/ImageOps.h]: Done.

File: libs/image/include/image/LinearImage.h (hunk 4 above)
Comment (prideout) [reply libs/image/include/image/LinearImage.h]: Done.

File: libs/image/src/ImageSampler.cpp (hunk 5 above)
Comment (prideout) [reply libs/image/src/ImageSampler.cpp]: Done.

File: libs/image/include/image/ColorTransform.h (hunk 2 above)
Comment (prideout) [reply libs/image/include/image/ColorTransform.h]: This was copied from old code but I added comments and assertions to clarify the input / output expectations for these functions.

**Pull request merged**